Effect Hook page (#117)
reactjs/pt-BR.reactjs.org: content/docs/hooks-effect.md

---
id: hooks-state
title: Usando Effect Hook (Hook de Efeito)
permalink: docs/hooks-effect.html
next: hooks-rules.html
prev: hooks-intro.html
---

_Hooks_ são uma nova adição ao React 16.8. Eles permitem que você use o state e outros recursos do React sem escrever uma classe.

O *Effect Hook* (Hook de Efeito) te permite executar efeitos colaterais em componentes funcionais:

```js{1,6-10}
import React, { useState, useEffect } from 'react';

> function Exemple() {

Reviewer: ```suggestion
function Exemplo() {
```

  const [count, setCount] = useState(0);

  // Similar ao componentDidMount e componentDidUpdate:
  useEffect(() => {

>     // Atualiza o titulod do documento usando a API do browser

Author: arrumar titulo

Existem dois tipos comuns de efeitos colaterais nos componentes React: aqueles que não precisam de limpeza, e aqueles que precisam. Vamos ver as suas diferenças mais detalhadamente.

## Efeitos Sem Limpeza {#effects-without-cleanup}

De vez em quando, nós queremos **executar algum código adicional depois que o React atualizou a DOM.** Requisições, mutações manuais do DOM e log são exemplos comuns de efeitos que não precisam de limpeza. Nós dizemos isso porque podemos executa-los e imediatamente esquecer deles. Vamos comparar como classes e Hooks nos permitem expressar tais efeitos colaterais.

### Exemplo Usando Classes {#example-using-classes}

Em um componente de classe do React, o método `render` não deve causar efeitos colaterais. Seria muito cedo -- nós geralmente queremos executar nossos efeitos *depois* que o React atualizou a DOM.

Isso é o porque nas classes do React, nós colocamos efeitos dentro de `componentDidMount` e `componentDidUpdate`. Voltando ao nosso exemplo, aqui está um componente de classe do React chamado contador que atualiza o titulo logo após o React faz as mudanças na DOM:

```js{9-15}

> class Exemple extends React.Component {

Reviewer: Use `Example` ou `Exemplo`
```suggestion
class Exemplo extends React.Component {
```

  const [count, setCount] = useState(0);

  useEffect(() => {
    document.title = `Você clicou ${count} vezes`;
  });

  return (
    <div>
      <p>Você clicou {count} vezes</p>
      <button onClick={() => setCount(count + 1)}>
        Clique aqui
      </button>
    </div>
  );
}
```

> **O que o `useEffect` faz?** Usando esse Hook, você diz ao React que o componente precisa fazer algo apenas depois da renderização. O React ira se lembrar da função que voce passou (nos referiremos a ele como nosso "efeito"), e chamá-la depois que realizar as atualizações do DOM. Nesse efeito, nós mudamos o título do documento, mas nós podemos também realizar busca de dados ou chamar alguma API imperativa.

Reviewer: ```suggestion
**O que o `useEffect` faz?** Usando esse Hook, você diz ao React que o componente precisa fazer algo apenas depois da renderização. O React ira se lembrar da função que você passou (nos referiremos a ele como nosso "efeito"), e chamá-la depois que realizar as atualizações do DOM. Nesse efeito, mudamos o título do documento, mas podemos também realizar busca de dados ou chamar alguma API imperativa.
```

**Porque `useEffect` é chamado dentro de um componente?** Colocando `useEffect` dentro do componente nos permite acessar a variavel de estado `count` (ou qualquer outra prop) direto do efeito. Nós não precisamos de uma API especial para lê-los -- já esta no escopo da função. Hooks adota as closures do JavaScript e evita API's especificas do React onde o JavaScript já provê uma solução.

> **`useEffect` executa depois de toda renderização?** Sim! Por padrão, ele roda depois da primeira renderização *e* depois de toda atualização. (Falaremos sobre [como customizar isso](#tip-optimizing-performance-by-skipping-effects) depois.) Ao invés de pensar em temos de "montando" e "atualizando", você pode achar mais facil pensar que efeitos acontecem "depois da renderização". React garante que o DOM foi atualizado na hora de executar os efeitos.

Author: Ali em 'temos' deveria ser `termos`

### Explicação detalhada{#detailed-explanation}

Agora que já sabemos mais sobre os efeitos, essas linhas devem fazer sentido:

```js
function Example() {
  const [count, setCount] = useState(0);

  useEffect(() => {
    document.title = `Você clicou ${count} vezes`;
  });
```

> Nós declaramos o state `count`, e então dizemos ao React que precisamos usar um efeito. Nós passamos uma função para o Hook `useEffect`. Essa função que nós passamos *é* o nosso efeito. Dentro do nosso efeito, nós definimos o titulo do documento usando `document.title` da API do navegador. Nós podemos ler o último `count` dentro do nosso efeito por que ele está dentro do escopo da nossa função. Quando o React renderizar nosso componente, ele ira se lembrar do efeito que usamos, e então executar os nossos efeitos depois de atualizar o DOM. Isso acontece para cada renderização, incluindo a primeira.

Reviewer: ```suggestion
Declaramos o state `count`, e então dizemos ao React que precisamos usar um efeito. Passamos uma função para o Hook `useEffect`. Essa função que passamos *é* o nosso efeito. Dentro do nosso efeito, definimos o título do documento usando `document.title` da API do navegador. Podemos ler o último `count` dentro do nosso efeito por que ele está dentro do escopo da nossa função. Quando o React renderizar nosso componente, ele ira se lembrar do efeito que usamos, e então executar os nossos efeitos depois de atualizar o DOM. Isso acontece para cada renderização, incluindo a primeira.
```

> Desenvolvedores JavaScript experientes podem perceber que a função passada para o `useEffect` vai ser diferente a cada renderização. Isso é intencional. Na verdade, isso é o que nos deixa ler o valor de `count` de dentro do efeito sem nos preocuparmos com ele ficar obsoleto. Toda vez que nós re-renderizarmos, nós agendamos um efeito _diferente_, substituindo o antigo. De uma maneira, isso faz os efeitos se comportarem mais como o resultado da renderização -- cada efeito "pertence" à sua renderização especifica. Nós vamos ver mais claramente como isso pode ser util [depois, nessa página](#explanation-why-effects-run-on-each-update).

Reviewer: ```suggestion
Desenvolvedores JavaScript experientes podem perceber que a função passada para o `useEffect` vai ser diferente a cada renderização. Isso é intencional. Na verdade, isso é o que nos deixa ler o valor de `count` de dentro do efeito sem nos preocuparmos com ele ficar obsoleto. Toda vez que nós re-renderizarmos, agendamos um efeito _diferente_, substituindo o antigo. De uma maneira, isso faz os efeitos se comportarem mais como o resultado da renderização -- cada efeito "pertence" à sua renderização especifica. Vamos ver mais claramente como isso pode ser útil [depois, nessa página](#explanation-why-effects-run-on-each-update).
```

  });
```

Outros efeitos não precisam ter a fase de limpeza, e não retornam nada.

```js
  useEffect(() => {
    document.title = `Você clicou ${count} vezes`;
  });
```

> O Hook de Efeito unifica ambos casos com uma unica API.

Reviewer 2: ```suggestion
O Effect Hook unifica ambos casos com uma única API.
```

-------------

> **Se você acha que tem um entendimento decente de como o Hook de Efeito funciona, ou se você se sente sobrecarregado, você já pode pular para a [proxima pagina sobre as Regras dos Hooks](/docs/hooks-rules.html).**

Reviewer 2: ```suggestion
**Se você acha que tem um bom entendimento de como o Effect Hook funciona, ou se você se sente sobrecarregado, você já pode pular para a [próxima pagina sobre as Regras dos Hooks](/docs/hooks-rules.html).**
```

Esse "decente" aqui soa um pouco pejorativo, creio que a melhor tradução nesse caso não é a literal.

-------------

## Dicas para Usar Efeitos {#tips-for-using-effects}

> Nós vamos continuar essa página com um olhar mais aprofundado em alguns aspectos do `useEffect` sobre os quais alguns usuários mais experientes do React provavelmente ficarão curiosos. Não se sinta obrigado a se aprofundar neles agora. Você sempre pode voltar nessa página para saber mais sobe o Effect Hook.

Reviewer: ```suggestion
Vamos continuar essa página com um olhar mais aprofundado em alguns aspectos do `useEffect` sobre os quais alguns usuários mais experientes do React provavelmente ficarão curiosos. Não se sinta obrigado a se aprofundar neles agora. Você sempre pode voltar nessa página para saber mais sobe o Effect Hook.
```

Quando nós renderizamos com `count` atualizado para `6`, o React irá comparar os itenst na array `[5]` da renderização passada com os items no array `[6]` da próxima renderização. Desta vez, o React irá re-aplicar o efeito porque `5 !== 6`. Se houver multiplos itens na array, o React ira re-executar o efeitos mesmo se apenas um deles for diferente.

Isso também funciona para efeitos que tenham uma fase de limpeza:

```js{6}
useEffect(() => {
  ChatAPI.subscribeToFriendStatus(props.friend.id, handleStatusChange);
  return () => {
    ChatAPI.unsubscribeFromFriendStatus(props.friend.id, handleStatusChange);
  };
}, [props.friend.id]); // Apenas re-subscribe se props.friend.id mudar
```

No futuro, talvez o segundo argumento seja adicionado automaticamente por uma transformação em tempo de build.

>Nota
>
>Se você usar essa otimização, tenha certeza de que a array inclua **qualquer valor do escopo acima que mude com o tempo e que ele seja usado pelo efeito**. Do contrario, seu código ira referenciar valores estagnados da renderização passada. Nós iremos discurtir outras opções de otimização nas [referencias da API do Hooks](/docs/hooks-reference.html).
>

> >Se você quer executar um efeito e limpa-lo apenas uma vez (na montagem e desmontagem), você pode passar uma array vazia (`[]`) como segundo argumento. Isso diz ao React que o seu efeito não depende de *nenhum* valor das props ou state, então ele nunca precisa re-executar. Isso não é tratado como um caso especial -- ele segue diretamente como os inputs de uma array sempre funcionaram. Passando `[]` fica mais perto do familiar modelo mental de `componentDidMount` e `componentWillUnmount`, contudo nós não sugerimos fazer disso um habito pois facilmente gera bugs, [como discutido acima](#explanation-why-effects-run-on-each-update). Não esqueca de que o React adia a execução do `useEffect` até o browser ser pintado, então fazer trabalho extra é menos problematico.

Reviewer 2: ```suggestion
>Se você quer executar um efeito e limpá-lo apenas uma vez (na montagem e desmontagem), você pode passar um array vazio (`[]`) como segundo argumento. Isso diz ao React que o seu efeito não depende de *nenhum* valor das props ou state, então ele nunca precisa re-executar. Isso não é tratado como um caso especial -- ele segue diretamente a maneira como os arrays passados como input do Hook sempre funcionam. Passar `[]` aproxima-se do modelo mental familiar de `componentDidMount` e `componentWillUnmount`, contudo nós sugerimos não fazer disso um hábito pois geralmente gera bugs, [como discutido acima](#explanation-why-effects-run-on-each-update). Não esqueça de que o React adia a execução do `useEffect` até o navegador ser pintado, então fazer trabalho extra é menos problemático.
```

## Próximos Passos {#next-steps}

> Parabens! Essa foi uma página longa, mas com sorte ao fim a maioria das suas perguntas sobre efeitos foram respondidas. Você aprendeu sobre o Hook State e o Hook Effect, e ainda hà *muito* mais o que você pode fazer com os dois combinados. Eles cobrem a maioria dos casos de uso para classes -- e ainda algumas coisas que elas não fazem, talvez você encontre algums [Hooks adicionais](/docs/hooks-reference.html) uteis.

Reviewer 2: ```suggestion
Parabéns! Essa foi uma página longa, mas com sorte ao fim a maioria das suas perguntas sobre efeitos foram respondidas. Você aprendeu sobre o Hook State e o Hook Effect, e há *muito* o que você pode fazer com os dois combinados. Eles cobrem a maioria dos casos de uso para classes -- e para os que eles não cobrirem, talvez você encontre alguns [Hooks adicionais](/docs/hooks-reference.html) úteis.
```

A semântica dessa última frase estava trocada. O texto se referia aos casos em que os Hooks **não** cobrissem os casos de uso para as classes.

> Nós tambem estamos começando a ver como Hooks resolvem problemas levantados na [Motivação](/docs/hooks-intro.html#motivation). Nós vemos como a limpeza dos efeitos evitam duplicação de código no `componentDidUpdate` e `componentWillUnmount`, mantém códigos relacionados juntos, e ajuda a evitar bugs. Nós tambem vimos como separar efeitos pelo seu proposito, que é uma coisa que não conseguiamos fazer com classes.

Reviewer 2: ```suggestion
Nós também estamos começando a ver como Hooks resolvem problemas levantados na [Motivação](/docs/hooks-intro.html#motivation). Nós vemos como a limpeza dos efeitos evitam duplicação de código no `componentDidUpdate` e `componentWillUnmount`, mantém códigos relacionados juntos, e ajuda a evitar bugs. Nós também vimos como separar efeitos pelo seu propósito, que é uma coisa que não conseguíamos fazer com classes.
```

> Nesse ponto você pode estar se perguntando como Hooks funcionam. Como o React sabe qual chamada do `useState` corresponde a qual vairavel do state entre as re-renderizações? Como o React "compara" os efeitos anteriores e os próximos toda atualiação? **Na próxima pagina nos iremos aprender sobre as [Regras dos Hooks](/docs/hooks-rules.html) -- elas são essenciais para fazer os Hooks funcionarem.**

Reviewer 2: ```suggestion
Nesse ponto você pode estar se perguntando como Hooks funcionam. Como o React sabe qual chamada do `useState` corresponde a qual variável de state entre as re-renderizações? Como o React "compara" os efeitos anteriores e os próximos toda atualização? **Na próxima pagina nos iremos aprender sobre as [Regras dos Hooks](/docs/hooks-rules.html) -- elas são essenciais para fazer os Hooks funcionarem.**
```
